### PR TITLE
Generate declaration files

### PR DESCRIPTION
**Pull request by Rich-Harris**
Another run at #9883. We can't use .js files as the source of truth inside a package in `node_modules`, but we _can_ use them as the source of truth for the `.d.ts` files. Combined with the `declarationMap` option, this makes 'go to definition' work as expected, and means that we don't have interfaces and documentation split between `.d.ts` and `.js` files.

This is WIP but I'm feeling fairly good about it. Downsides:

* we need to re-run `tsc` if we change interfaces while developing (but we still get HMR etc for non-interface changes, which is most of them)
* ~~we need to put types that can't be expressed in JSDoc in `.ts` files, but in order to retain the benefits of shipping raw source code, we need to ensure that we only export types, _not_ values, from those files. It might be worth having a build time check to guarantee that~~

---

## Update

Ok, so this mostly works (though the logic for generating docs from types needs to be updated). Summary:

* documentation and function interfaces etc live in source `.js` files
* we run `tsc` to turn `src/**/*.js` into `types/**/*.d.ts` (and `*.d.ts.map`, so 'go to definition' goes to the source)
* we copy over `src/**/*.d.ts` to `types`
* we adjust all the `import('types')` stuff to use relative imports, otherwise types will be broken for people not using `moduleResolution: 'bundler'`
* `pkg.exports` points at the source files, as before, but the `.` export additionally has a `types` condition which points at `types/exports/types.d.ts` (copied from `src/exports/types.d.ts`)
* `pkg.types` and `pkg.typesVersions` are configured so that all this continues to work without `moduleResolution: 'bundler'`

Trade-offs:

* VSCode will no longer autocomplete `@sveltejs/kit/*`, only `@sveltejs/kit` itself. This is rather annoying. I'm not sure if there's a piece missing that would fix this. We _could_ generate the `declare module` declarations from the `.d.ts` files, I suppose, though that would kinda suck
* This adds a prepublish step where there wasn't one before. I think this trade-off is worthwhile, but it deserves to be called out

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ ] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Review by dummdidumm — COMMENTED** [review]
Using the JSdoc output without further cleanup looks like this on hover:
image
which is not very pleasing an the `@template` is straight up confusing for someone not familiar with it. This function makes it nicer:
```js
function adjust(input) {
	// Extract the import paths and types
	const import_regex = /import\(("|')(.+?)("|')\)\.(\w+)/g;
	let import_match;
	const import_map = new Map();
  
	while ((import_match = import_regex.exec(input)) !== null) {
		const imports = import_map.get(import_match[2]) || new Set();
		imports.add(import_match[4]);
		import_map.set(import_match[2], imports);
	}
  
	// Replace inline imports with their type names
	const transformed = input.replace(import_regex, "$4");
  
	// Remove/adjust @template, @param and @returns lines
	const lines = transformed.split("\n");
	const filtered_lines = lines.filter(line => {
		line = line.trim();
	  	return !line.startsWith("* @template") &&
			!((line.startsWith("* @returns") || line.startsWith("* @param")) && line.endsWith("}"));
	}).map(line => {
		line = line.trim();
		// Strip {...} from @param and @returns lines
		if (line.startsWith("* @param {") || line.startsWith("* @returns {")) {
			let openCount = 1;
			let closedCount = 0;
			let i = line.indexOf('{') + 1;
			for (; i < line.length; i++) {
				if (line[i] === "{") openCount++;
				if (line[i] === "}") closedCount++;
				if (openCount === closedCount) break;
			}
			line = line.slice(0, line.indexOf('{')) + line.slice(i + 1);
		}
		return line;
	});
  
	// Replace generic type names with their plain versions
	const renamed_generics = filtered_lines.map(line => {
	  return line.replace(/(\W|\s)([A-Z][\w\d$]*)_\d+(\W|\s)/g, "$1$2$3");
	});
  
	// Generate the import statement for the types used
	const import_statements = Array.from(import_map.entries())
	  .map(([path, types]) => `import { ${[...types].join(', ')} } from '${path}';`)
	  .join("\n");
  
	return [import_statements, ...renamed_generics].join("\n");
}
```

File: packages/kit/src/exports/index.js [hunk 1]
```diff
@@ -25,7 +31,11 @@ export function redirect(status, location) {
 	return new Redirect(status, location);
 }
 
-/** @type {import('@sveltejs/kit').json} */
+/**
+ * Create a JSON `Response` object from the supplied data.
+ * @param {any} data The value that will be serialized as JSON.
+ * @param {ResponseInit} [init] Options such as `status` and `headers` that will be added to the response. `Content-Type: application/json` and `Content-Length` headers will be added automatically.
+ */
```
[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/kit/src/exports/index.js:38]: Lol wtf why did Github show that comment, I thought I deleted it. Yes there's no generic

File: packages/kit/src/exports/index.js [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
 /**
- * @type {import('@sveltejs/kit').error}
- * @param {number} status
- * @param {any} message
+ * Creates an `HttpError` object with an HTTP status code and an optional message.
+ * This object, if thrown during request handling, will cause SvelteKit to
+ * return an error response without invoking `handleError`.
+ * Make sure you're not catching the thrown error, which would prevent SvelteKit from handling it.
+ * @param {number} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses). Must be in the range 400-599.
+ * @param {{ message: string } extends App.Error ? App.Error | string | undefined : never} message An object that conforms to the App.Error type. If a string is passed, it will be used as the message property.
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/exports/index.js:24]: added `@overload`

[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/exports/index.js:24]: ah nvm figured it out, it was unrelated — i had removed `.ts` files from `tsconfig.build.json`

[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/kit/src/exports/index.js:24]: You need the overload because the type definition is:
- if you did not modify `App.Error`, in other words only `{ message: string }` is required, then the second parameter is optional (because it is falling back to `Error: ${status}` in that case)
- if you did modify `App.Error`, in other words more than just `{ message: string }` is requreid, then you can't omit the second parameter and is required

You can only define this with overloads

File: packages/kit/src/exports/index.js [hunk 3]
```diff
@@ -65,9 +93,11 @@ export function text(body, init) {
...
 
 /**
- * Generates an `ActionFailure` object.
- * @param {number} status
- * @param {Record<string, any> | undefined} [data]
+ * Create an `ActionFailure` object.
+ * @template {Record<string, any> | undefined} T
+ * @param {number} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses). Must be in the range 400-599.
+ * @param {T} [data] Data associated with the failure (e.g. validation errors)
+ * @returns {ActionFailure<T>}
```
Comment (Rich-Harris) [inline packages/kit/src/exports/index.js:100]: There's something about this change that's causing one test to fail — the `ActionResult` here...

https://github.com/sveltejs/kit/blob/f344267982002dad40f83b20f05ced5c787c6325/packages/kit/src/core/sync/write_types/test/actions/%2Bpage.server.js#L41

...has this type:

```ts
ActionResult<{
    status: number;
    data: {
        fail: string;
    } | undefined;
} | {
    success: boolean;
} | {
    id: number;
    username: string;
    profession: string;
} | {
    status: number;
    data: {
        reason: {
            error: {
                code: "VALIDATION_FAILED";
            };
        };
    } | undefined;
} | {
    ...;
}, never>
```

It should have this type:

```ts
ActionResult<{
    success: boolean;
} | {
    id: number;
    username: string;
    profession: string;
}, {
    fail: string;
} | {
    reason: {
        error: {
            code: "VALIDATION_FAILED";
        };
    };
}>
```

In other words, the `ExcludeActionFailure` and `ExtractActionFailure` types in the generated `$types.d.ts` are misfiring:

```ts
type ExcludeActionFailure<T> = T extends Kit.ActionFailure<any> ? never : T extends void ? never : T;
type ActionsSuccess<T extends Record<string, (...args: any) => any>> = { [Key in keyof T]: ExcludeActionFailure<Awaited<ReturnType<T[Key]>>>; }[keyof T];
type ExtractActionFailure<T> = T extends Kit.ActionFailure<infer X>	? X extends void ? never : X : never;
type ActionsFailure<T extends Record<string, (...args: any) => any>> = { [Key in keyof T]: Exclude<ExtractActionFailure<Awaited<ReturnType<T[Key]>>>, void>; }[keyof T];
type ActionsExport = typeof import('../../../../../../../../+page.server.js').actions
export type SubmitFunction = Kit.SubmitFunction<Expand<ActionsSuccess<ActionsExport>>, Expand<ActionsFailure<ActionsExport>>>
```

Instead of getting a `SubmitFunction<Success, Failure>` we're getting something like a `SubmitFunction<Success & Failure, never>`. I'm not sure why.

File: packages/kit/src/exports/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/exports/index.js:24]: gotcha, i missed the fact that one was optional and one wasn't. updated

File: packages/kit/src/exports/index.js [hunk 4]
```diff
@@ -65,9 +79,9 @@ export function text(body, init) {
 }
 
 /**
- * Generates an `ActionFailure` object.
- * @param {number} status
- * @param {Record<string, any> | undefined} [data]
+ * Create an `ActionFailure` object.
+ * @param {number} status The [HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses). Must be in the range 400-599.
+ * @param {Record<string, any> | undefined} [data] Data associated with the failure (e.g. validation errors)
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/exports/index.js]: updated, but i think i screwed it up somehow: https://github.com/sveltejs/kit/pull/9895#discussion_r1190329703

File: packages/kit/src/types/public.d.ts [hunk 5]
```diff
@@ -18,8 +16,9 @@ import {
 	RequestOptions,
 	RouteSegment,
 	UniqueInterface
-} from './private.js';
+} from './private';
```
Comment (dummdidumm) [inline packages/kit/src/types/public.d.ts:19]: AFAIK all type imports need the `.js` ending or else `moduleResolution: node16/nodenext` will error. It's stupid, but it is what it is.

File: packages/kit/src/types/public.d.ts [hunk 6]
```diff
@@ -1053,8 +1052,8 @@ export interface RouteDefinition<Config = any> {
 	config: Config;
 }
 
-export class Server {
-	constructor(manifest: SSRManifest);
+export interface Server {
+	new (manifest: SSRManifest): Server;
```
Comment (dummdidumm) [inline packages/kit/src/types/public.d.ts:1056]: Why this change? Curious if these are equivalent.

File: packages/kit/src/exports/index.js [hunk 7]
```diff
@@ -65,9 +93,11 @@ export function text(body, init) {
 }
 
 /**
- * Generates an `ActionFailure` object.
- * @param {number} status
- * @param {Record<string, any> | undefined} [data]
+ * Create an `ActionFailure` object.
+ * @template {Record<string, any> | undefined} T
```
Comment (dummdidumm) [inline packages/kit/src/exports/index.js:97]: This defaults the `T` to `undefined` if not set which makes it equivalent to the `d.ts` definition. Maybe that fixes the bug you asked about three lines below.
```suggestion
 * @template {Record<string, any> | undefined} [T=undefined]
```

**Review by dummdidumm — COMMENTED** [review]
One way to solve the auto completion problem would be to not use `typesVersions` and instead create `d.ts` files which TypeScript will naturally pick up:
```
kit
  hooks.d.ts -> just reexports everything from types/../hooks/index.ts or wherever
  vite.d.ts -> same
  node/index.d.ts -> same
  node/polyfills.d.ts -> same
```

This pollutes the project a bit but I'm willing to take that hit for better DX.

In general my gripe with this change is that things are now split up. Interfaces are at X, methods at Y, it's all a bit scattered - maybe I'm just not used to it. This is the one thing TS makes better IMO, you can colocate this stuff better.

File: packages/kit/src/exports/index.js (hunk 7 above)
Comment (dummdidumm) [reply packages/kit/src/exports/index.js]: Mhm doesn't seem to have changed the test failure outcome, but the change is still correct.

File: packages/kit/src/types/public.d.ts (hunk 6 above)
Comment (Rich-Harris) [reply packages/kit/src/types/public.d.ts:1056]: ah this is because at one point i was experimenting with having `.ts` files in `src` instead of `.d.ts` files, so that all the `.d.ts` files would be generated. (i think having `.d.ts` files in `src` is better, on reflection — we just need to copy them over to the `types` directory). you can't declare a `class` in a `.ts` file without actually implementing it

now that it's in a `.d.ts` again we could revert

**Comment by Rich-Harris** [comment]
> One way to solve the auto completion problem would be to not use `typesVersions` and instead create `d.ts` files which TypeScript will naturally pick up:

Unfortunately VSCode is liable to create broken imports like this...

```js
import { thing } from '@scope/package/subpackage/index.js';
```

...when you want this:

```js
import { thing } from '@scope/package/subpackage';
```

It's dependent on the user's config etc, so it's unreliable.

> In general my gripe with this change is that things are now split up. Interfaces are at X, methods at Y, it's all a bit scattered - maybe I'm just not used to it. This is the one thing TS makes better IMO, you can colocate this stuff better.

TS definitely is better on this dimension, though I actually think this PR is better than what we currently have — all the types are in `.d.ts` files, all the values are in `.js` files, instead of having types _and_ values in `.d.ts` files _and_ having values in `.js` files. It feels like a cleaner separation to me.

Though it is a shame that we have to use `declare module` for all the `$app` stuff, and export functions and stuff from there. Would love to figure out a way to drive that from source (not least for 'go to definition'), but I haven't figured out what that could look like

**Comment by Rich-Harris** [comment]
After playing around a bit more I've come to the conclusion that what we really want is to generate (from source, with sourcemaps) a single `.d.ts` that bundles all the types and uses `declare module`:

```ts
declare namespace App {
  export interface Error {...}
}

declare module '@sveltejs/kit' {
  export interface Adapter {...}
  export function error(status: number, body: App.Error): ...
  // etc
}

declare module '@sveltejs/kit/hooks' {
  import { Handle } from '@sveltejs/kit';
  export function sequence(...handlers: Handle[]): Handle;
}

declare module '$app/environment' {
  export const browser: boolean;
  // ...
}

// etc

//# sourceMappingURL=index.d.ts.map
```

The `index.d.ts.map` file would map back to the original source.

This is the most reliable way to get auto-imports working across all TypeScript configurations, and is the _only_ way to make it work with modules like `$app/environment`, since they can't be expressed in a `pkg.exports` map. It's basically what we have now, except generated from source instead of handwritten, which would a) simplify the codebase and b) allow us to map back to the original source. Additionally, having everything in a single file ought to speed up startup (i.e. less time looking at a 'loading...' indicator while the TS server figures out where everything is).

Unfortunately, a tool for creating such a `.d.ts` bundle doesn't currently exist. There's a couple of tools that are _related_ (api-extractor, rollup-plugin-dts, dts-bundle) but none of them are really suitable for this use case. I think such a tool _should_ exist and I wouldn't mind having a go at building it, but I can't pretend it's a top priority.

**Comment by Rich-Harris** [comment]
going to close this and maybe take another run at it later
